src/pyk/proof/proof.py

```diff
@@ -175,9 +176,52 @@ def json(self) -> str:
         return json.dumps(self.dict)
 
     @property
-    def summary(self) -> Iterable[str]:
+    def summary(self) -> ProofSummary:
+        @dataclass
+        class _BaseSummary(ProofSummary):
+            id: str
+            status: ProofStatus
+
+            @property
+            def lines(self) -> list[str]:
+                return [f'Proof: {self.id}', f'    status: {self.status}']
+
         subproofs_summaries = [subproof.summary for subproof in self.subproofs]
-        return chain([f'Proof: {self.id}', f'    status: {self.status}'], *subproofs_summaries)
+        return CompositeSummary([_BaseSummary(self.id, self.status), *subproofs_summaries])
+
+
+@dataclass
+class ProofSummary(ABC):
+    id: str
+    status: ProofStatus
+
+    @property
+    @abstractmethod
+    def lines(self) -> list[str]:
+        return [f'Proof: {self.id}', f'    status: {self.status}']
```

Comment:
Let's keep this class proper abstract.

```suggestion
class ProofSummary(ABC):
    @property
    @abstractmethod
    def lines(self) -> list[str]:
        ...
```